src/Core/Auth/UserFeatures/PremiumAccess/PremiumAccessQuery.cs

```diff
@@ -0,0 +1,97 @@
+    private readonly IApplicationCacheService _applicationCacheService;
+
+    public PremiumAccessQuery(
+        IOrganizationUserRepository organizationUserRepository,
+        IApplicationCacheService applicationCacheService)
+    {
+        _organizationUserRepository = organizationUserRepository;
+        _applicationCacheService = applicationCacheService;
+    }
+
+    public async Task<bool> CanAccessPremiumAsync(User user)
+    {
+        return await CanAccessPremiumAsync(user.Id, user.Premium);
+    }
+
+    public async Task<bool> CanAccessPremiumAsync(Guid userId, bool hasPersonalPremium)
+    {
+        if (hasPersonalPremium)
+        {
+            return true;
+        }
+
+        return await HasPremiumFromOrganizationAsync(userId);
+    }
+
+    public async Task<bool> HasPremiumFromOrganizationAsync(Guid userId)
+    {
+        // Note: GetManyByUserAsync only returns Accepted and Confirmed status org users
+        var orgUsers = await _organizationUserRepository.GetManyByUserAsync(userId);
+        if (!orgUsers.Any())
+        {
+            return false;
+        }
+
+        var orgAbilities = await _applicationCacheService.GetOrganizationAbilitiesAsync();
+        return orgUsers.Any(ou =>
+            orgAbilities.TryGetValue(ou.OrganizationId, out var orgAbility) &&
+            orgAbility.UsersGetPremium &&
+            orgAbility.Enabled);
+    }
+
+    public async Task<Dictionary<Guid, bool>> CanAccessPremiumBulkAsync(IEnumerable<User> users)
+    {
+        var result = new Dictionary<Guid, bool>();
+        var usersList = users.ToList();
+
+        if (!usersList.Any())
+        {
+            return result;
+        }
+
+        var userIds = usersList.Select(u => u.Id).ToList();
+
+        // Get all org memberships for these users in one query
+        // Note: GetManyByManyUsersAsync only returns Accepted and Confirmed status org users
+        var allOrgUsers = await _organizationUserRepository.GetManyByManyUsersAsync(userIds);
+        var orgUsersGrouped = allOrgUsers
+            .Where(ou => ou.UserId.HasValue)
+            .GroupBy(ou => ou.UserId!.Value)
+            .ToDictionary(g => g.Key, g => g.ToList());
+
+        var orgAbilities = await _applicationCacheService.GetOrganizationAbilitiesAsync();
+
+        foreach (var user in usersList)
+        {
+            var hasPersonalPremium = user.Premium;
+            if (hasPersonalPremium)
+            {
+                result[user.Id] = true;
+                continue;
+            }
+
+            var hasPremiumFromOrg = orgUsersGrouped.TryGetValue(user.Id, out var userOrgs) &&
+                userOrgs.Any(ou =>
+                    orgAbilities.TryGetValue(ou.OrganizationId, out var orgAbility) &&
+                    orgAbility.UsersGetPremium &&
+                    orgAbility.Enabled);
+
+            result[user.Id] = hasPremiumFromOrg;
+        }
+
+        return result;
+    }
```

Comment:
This has a database call + cache + a nested loop to get information that ultimately comes from the database anyway.

I suggest a simpler implementation would be a new sproc for this purpose (untested):

```sql
    SELECT
        U.[Id],
        MAX(U.[Premium]) PersonalPremium,
        MAX(O.[UsersGetPremium]) OrganizationPremium
    FROM
        [dbo].[UserView] U
        JOIN [dbo].[OrganizationUserView] OU ON U.Id = OU.UserId
        JOIN [dbo].[Organization] O ON OU.OrganizationId = O.Id
    WHERE
        U.[Id] IN (SELECT [Id] FROM @Ids)
    GROUP BY
        U.[Id]
```

And then this code could just return `PersonalPremium || OrganizationPremium` for each user.

Reply:
I'm using the existing `UserWithCalculatedPremium` for now. I'll refactor that later here on in a separate branch